### ISSUE TITLE
Bug: share/vss/rabin MinimumT(6)=3 instead of 4

### ISSUE THREAD

**Issue by vibs29** [dedis/kyber#474]
https://github.com/dedis/kyber/blob/master/share/vss/rabin/vss.go#L697
Should that be (n/2)+1 instead of (n+1)/2?

**Comment by ineiti** [comment]
If you look at subsection 4.4 of the [paper](http://nozdr.ru/data/media/biblio/kolxoz/Cs/CsLn/Information%20Security%20and%20Privacy,%206%20conf.,%20ACISP%202001(LNCS2119,%20Springer,%202001)(ISBN%203540423001)(531s).pdf#page=426 ), you see that it says:

```
We have to show that if less than t players are corrupted, 
the scheme always produces a valid signature. 
We assume that t ≤ n/2 .
```

So I made a small table and calculated it by hand with your `(n/2)+1` and compared it with the original `(n+1)/2`, and the original fits the paper, while your proposal does not.

But I might be wrong :) So please verify with the paper and tell me if after all you can accept that `(n+1)/2` is effectively correct...

**Comment by vibs29** [comment]
Hi Linus. Thanks for looking at this.
My patch was based mostly on how the Kyber poly_test.go used (N/2)+1
instead of (N+1)/2. (Integer division).
Your response told me I needed to study the theory carefully.
I have now and am glad. I was more wrong than I thought.
In fact, the flaw in the library is worse!
The direction of the inequality you quoted from the Stinson paper is
the clue.
It's the opposite of what's in Kyber.
The bound on T should be upper not lower.
Kyber mistakenly advises a lower bound instead of an upper bound.

(A) That's the biggest thing to fix by far. I'll provide an intuition
below for why the paper places an upper-bound on T not a lower bound.

(B) Once that's fixed, an optional minor improvement is that you can place
the upper-bound a little higher than the Stinson paper says because it
doesn't (at end of section 2.4) quote the full range of T's that the
underlying DKG paper by Gennaro really proves usable. Gennaro really
allows[*] as high as T < N/2 + 1 (real division). Stinson's proof in
section 4.4 will remain correct even if applied up to this slightly
higher T.

====

(A) Why Stinson, Gennaro don't even really consider large T's and only
    bother proving their techniques for small T's:

The reason is that no threshold scheme can even in principle work for
large T's. Gennaro mentions (in his section 2.1) that T-1 < N/2 is the
best achievable for _any_ solution. This is why nobody bothers proving
anything for larger T's, because if you want a larger T, you've already
lost. No past, present or future technique for addressing this particular
adversarial model can possibly save you. This was so obvious to the
authors that they had no space in their margins to explain why! I'll try.

The Gennaro paper models the threat that an adversary (Eve) controls some
unknown subset consisting of T-1 players. I'll call players she controls
"puppets". To guard against Eve therefore, any technique including
Gennaro's _must_ prevent every set of T-1 players from reconstructing the
secret -- (1). Every set because it's not known which set is Eve's. That's
secrecy.

[Margin note: It's not worth thinking about T's so large that T-1=N
because that's just the situation that Eve controls all N players,
in which case no threshold scheme is applicable anyway.]

I'm making the case that not all values of T-1 are sensible. But I
won't claim that the requirement for secrecy alone places any meaningful
constraints on T-1. Indeed it likely places none. The constraint on T-1
will come from ideas described next.

--
consider N = 10, T - 1 = 5, i.e. there are 5 puppets.
Our mathematical technique requires 6 honest players for secret
reconstruction.
But (and this is the insight) every set of 6 players here necessarily
includes at least one puppet.
Via this puppet, Eve may be be uncooperative in e.g. secret
reconstruction.
Thus the overall group's success depends on the whim of Eve.
No threshold scheme of the kind that (1) above is can overcome this.
That's why the various authors don't even consider such high T's to be
in scope.

So    T-1;N = 5;10 is a doomed configuration.
So is T-1;N = 5;9
But the following are OK, not doomed:
T-1;N = 4;10
T-1;N = 4;9
A doomed configuration is one where every set of size T has been
infiltrated by Eve.
A not doomed configuration is one where there is at least one set of
size T that hasn't been infiltrated.

--
When the various papers say, "we assume T < ...", I interpret that not in
the sense of "so far we have only proven our technique safe for choices
of T that are this small" but rather in the sense of "we plan to show
our technique is safe for all choices of T that aren't otherwise doomed
due for the well-known reason".

In other words, they consider T in the doomed range as being out of
scope just as my margin note above made clear that T-1=N was out of
scope. Anyone coming to threshold schemes (whether to their papers or
their software libraries) with such large values of T is in the wrong
place to begin with. No threshold scheme can help a person who shows up
with such an absurd T. No threshold scheme can be safe for such large
T's, even in theory. That's why the papers say "we asume T < ...",
because why should they bother even thinking about larger T's? Surely
everybody's got the memo by now that large T's are doomed to begin with.

--
I described secrecy above (in (1)). There's also the idea of
robustness. Gennaro mentions that the theoretical upper bound on T-1
is for solutions that are both secret and robust. That means that
correct, secret solutions that work with larger T's necessarily lose
robustness. Robustness is some combination of the idea that T honest
players will always reconstruct the same secret and that infiltrators
will be detected. Neither Gennaro nor Stinson care about non-robust
solutions so the bound on T is a real one for them.

====

(B) The optional improvement:

Stinson (section 2.4), perhaps by mistake, didn't use the full range
that non-doomed T's can take and that Gennaro proves his scheme is
suitable for.
So he assumes T's will have a smaller upper-bound and proves his technique
for that restricted range. However, the only place he makes use of the
range is in section 4.4. And it's easy to see that 4.4 would be true in
the full wider range of non-doomed T's.

In Go code, with integer division, going strictly by the Stinson paper
as it is written, T's are OK when this is true:
func NotDoomed(T int, N int) bool { return T <= N/2 }
and more generally, as I said, are OK when this is true:
func NotDoomed2(T int, N int) bool { return T <= (N+1)/2 }
(There will never be a threshold scheme that can safely allow a larger
T than NotDoomed2.)

====
Fix:

Implement a function to test that T is bounded above instead of below.
Communicate to all past library users that those who followed the
  library's original recommendation are actually using doomed T's.
Enforce the bounds check (instead of making it just advisory) to be
  doubly sure that future users don't choose a doomed T.
Housekeeping:
- share/poly_test.go should change to use non-doomed T's
- the other DKG's, VSS's, threshold signature scheme implementations
  may also need similar fixing

Even better than leaving the library user to type in an inequality sign
when validating their thresholds is to name your validation function
something like this instead of MaximumT():
func IsAbsurd(threshold int) bool {...}
so users know they only need to ensure they get a false.
(or alternatively call it NotAbsurd()...)
[Update: I've used the name NotDoomed above. The names are all only
         illustrative of course.]

====
Notes:

Which of the non-doomed T's should a library user actually choose?
As T increases from zero towards half-N, you get the ability to withstand
more powerful adversaries but also require more of the honest players
to be non-dormant.

====
References:

[*]:
Gennaro, R., Jarecki, S., Krawczyk, H. et al. Secure Distributed Key
Generation for Discrete-Log Based Cryptosystems. J Cryptology 20, 51–83
(2007). https://doi.org/10.1007/s00145-006-0347-3
Section 2.1
Note: When reading Gennaro's paper don't assume that it means the same
things by the same mathematical symbols as Stinson's paper. But symbols I
use anywhere in this note always mean what they do in Stinson/Kyber.

**Comment by seratym** [comment]
Hey @vibs29 and @ineiti, interesting discussion! I would like to contribute my views on the matter.
tldr: MinimumT (incorrectly) computes the threshold assuming the worst case number of corrupted players. The name should be changed and calculation adjusted.

## What might have been the source of confusion
First of all, as noted by @vibs29 , the symbol t denotes different things.
In general, t-out-of-n sharing means that t shares are required to reconstruct a secret. The t here stands for threshold (of required secret shares).
In the paper by Gennaro et al., t appears multiple times. The first mention ("who may corrupt more than t servers during the lifetime of the secret", Section 1), denotes the number of corrupted parties. In section 2.2, t is the degree of the polynomial, as used for Shamir's SS. Here t+1 shares are required to reconstruct the polynomial and therefore, 'coincidentally', an adversary that controls up to at most t parties cannot learn the secret. 

In the paper by Stinton et al. t is also used with different meanings. In section 2.2 it used to denote the threshold  for recovering the secret, but in section 8.1 t is used for time. 
In section 4.4, the protocol is proven to be robust if t <= n/2, where less than t players are corrupted. As @vibs29 has shown, if t is larger than the number of uncorrupted parties, then the benign parties will not be able to recover the secret (without help from the adversary). 

## Why t can be larger than n/2

Let us use f for the number of corrupted parties, t the threshold required to reconstruct the secret and n for the number of parties in total. For proper execution of the algorithm we have 2 requirements: robustness and secrecy.
**In the worst case**, I am positive that t = ceil((n + 1)/2) and f = n-t = floor((n - 1)/2). Secrecy is maintained since f < t. Robustness is also maintained since the number of uncorrupted players (n-f) > t.

> When the various papers say, "we assume T < ...", I interpret that not in
> the sense of "so far we have only proven our technique safe for choices
> of T that are this small" but rather in the sense of "we plan to show
> our technique is safe for all choices of T that aren't otherwise doomed
> due to the well-known reason".
Here you assume that f = t-1. Note that in my worst case definition, for even n the difference between f and t is 2, not 1.  

If we use DKG we need to assume an f. If we have a system with 200 players, but we assume there will be never more than 3 players corrupt, what t do we need to choose? From my understanding, 3 < t <= 197 would all be safe. Secrecy is assured, as well as robustness.

Both increasing or decreasing t allows for less corrupted players. If you decrease t, less corrupted players are needed to break the secrecy. If you increase t, less players are needed to break the robustness. There is no secure minimum for t, but setting t to just over half of n will make the protocol suitable for the largest amount of potentially corrupted players.

## My recommendation
At [vss.go#L699](https://github.com/dedis/kyber/blob/e423d8ebf8cce09d0e1cf532d3c6849ea262a534/share/vss/rabin/vss.go#L699C1-L699C20), change the line to ``return n / 2 + 1 `` as originaly proposed by @vibs29. Mind that Golang floors integer division. I believe the same change can be made to pedersen/vss.go.

> returns the minimum safe T that is proven to be secure with this protocol. 
Remove this comment.

Change the function name to something like ``DefaultT``.

**Comment by ineiti** [comment]
If I understand correctly, it doesn't matter in fact:
- `defT = int((n+1)/2)` - current implementation
  - with `n=10`, `defT=5`, there can only be up to 4 corrupt nodes, else they can reconstruct the secret
  - with `n=11`, `defT=6`, there can only be up to 5 corrupt nodes, else they can disrupt robustness and secrecy
- `defT = int(n/2) + 1` - proposed change
  - with `n=10`, `defT=6`, there can only be up to 4 corrupt nodes, else they disrupt robustness
  - with `n=11`, `defT=6`, there can only be up to 5 corrupt nodes, else they can reconstruct the secret or disrupt robustness

So either way, the number of maximum corrupt nodes before we lose either robustness or secrecy is the same for both formulas. But the `int(n/2) + 1` has a little bit more complexity for even number of nodes. Which makes me propose that we don't change the current formula...

Or we add two thresholds:
- `SecrecyOverRobustness = int(n/2) + 1`
- `RobustnessOverSecrecy = int((n+1) / 2)`

But anyway we need to adjust the comment here to reflect what you wrote:
https://github.com/dedis/kyber/blob/master/share/vss/rabin/vss.go#L134

And changing the name of `MinimumT` to `DefaultT` is an API breaking change, and as such would only be possible with kyber v4.

@pierluca which makes me think we can remove the `bug` tag.

**Comment by seratym** [comment]
I do agree. ``defT = int(n/2) + 1``  is more optimal. I am personally against adding two thresholds; if the threshold is that sensitive the implementer should rely on his own implementation. Instead, reflecting the trade-off in will aid in choosing a secure T.

**Comment by ineiti** [comment]
OK, so the PR comes down to update the documentation:

https://github.com/dedis/kyber/blob/master/share/vss/rabin/vss.go#L134
and
https://github.com/dedis/kyber/blob/master/share/vss/rabin/vss.go#L693

Does that seem a correct conclusion?

**Comment by seratym** [comment]
I think the same lines should be changed in pedersen aswell: e.g. github.com/dedis/kyber/blob/master/share/vss/pedersen/vss.go#L118

**Comment by vibs29** [comment]
Hi @seratym,

Thanks for studying all that so carefully. I know this is late, but I
hope you don't mind if I try and put my thoughts down nonetheless.

>  Here you assume that f = t-1.

I don't view f and t as independent variables. I view threshold
cryptography[1] as defining t in the following way (my words):
```
  Let c be the number of potentially corrupted parties.
  Then require a threshold of c+1 parties to cooperate in signing.
  Call this t.
```
In other words, t for threshold is an _abbreviation_ for c+1.
There is no benefit to requiring c+2 cooperators.

I therefore think it's mistaken to think of:
- f and t as independent, or
- f = n - t

And because:
- I think of the potentially corrupted parties as being _defined_
  as t-1, and
- Gennaro says this t-1 must be less than n/2 for secure and robust
  solutions

therefore,
ceil((n+1)/2)
is an invalid value of c+1 for n=6, because c=3 is larger than allowed.

In programming terms, the only reason an expert will launch the program
with a setting of n=6, t=4 will be because they think they can then
withstand 3 corrupted parties, but Gennaro tells us they are fooling
themselves; no algorithm can withstand 3/6 corrupted parties. Kyber
should tell such experts that t=4 is invalid and maybe even exit rather
than telling them it is valid.

I haven't yet thought about about advisory functions named
e.g. DefaultT(), which are for novices not experts. What experts and
novices alike want for starters from a validating function (which a name
like MinimumT() implies it is) is an objective answer to whether they are
running in an invalid configuration. n=6, t=4 is a misconfiguration. Said
another way, n=6, c=3 is a misconfiguration; the c is too large.

> If we have a system with 200 players, but we assume there will be
> never more than 3 players corrupt, what t do we need to choose? From
> my understanding, 3 < t <= 197 would all be safe. Secrecy is assured,
> as well as robustness.

As per my view above, this is equivalent to saying that
3 < c+1 <= 197
would all be safe, which Gennaro (section 2.1 "The Adversary") contradicts
by saying that c < n/2 is the best achievable. So really
0 < t <= 100
are safe. t > 100, i.e. c+1 > 100, are all unsafe. (And this is why
MinimumT() is a bug. It's named the opposite of what it should be.)

----
I stand by everything I wrote before including calling the function
named MinimumT that returns approximately n/2 a bug because it misleads
programmers into thinking only a larger t is valid when it's really only
a smaller t that is valid. Said more clearly, it misleads programmers
into thinking that a large c is valid when really it's only a small c
that is valid. I particularly stand by my earlier "Fixes" section. (If
you want an additional advisory function for novices named DefaultT(),
that function can return the strongest value, (n+1)/2 (integer division).)

----
References:
[1]: https://en.wikipedia.org/wiki/Threshold_cryptosystem
[2]: An older discussion https://github.com/dedis/kyber/issues/374

**Comment by ineiti** [comment]
It also depends on you're adversarial model is: if it's honest-but-curious servers, then a `t > n/2` does make sense. This means that an adversary wanting to find the aggregate private key (not stopping the protocol, like in your calculation), would need to get the private shares of more servers.

Of course this does not help if you have fully byzantine servers, in which case your evaluation is correct.

We should put the disclaimer in https://pkg.go.dev/go.dedis.ch/kyber more prominently in the [README.md](/dedis/kyber/blob/master/README.md), like https://github.com/dedis/kyber/pull/495

**Comment by vibs29** [comment]
Hi @ineiti. Here's how I see this.

1. My (i.e. some application writer's) adversarial model is irrelevant. The adversarial model assumed by _a library_ implementing a paper should be no weaker than that assumed by the paper.
2. Also, the library's validating function tells programmers that smaller than half is invalid and larger than half is valid whereas if anything it's smaller than half that's valid. That's a bug in the library that's more serious than an off-by-one error. Or even worse, it's a security flaw, in that it may have encouraged at least one non-expert to assume in their application program that they could withstand an impossibly high fraction of the worst kind of adversaries by setting a high T.

**Comment by seratym** [comment]
Hi @vibs29 and @ineiti, I hope you're doing well.
I am positive that t in Gennaro's paper stands for the number of adversaries, not the amount of shares needed to reconstruct the secret. Theorem 1's requirement for t is then a requirement for the number of adversaries, not the number of shares. T in kyber stands for the amount of shares needed for reconstruction. This does make the discussion confusion, at least for me.

minimumT does not make sense, maximumT doesn't either. Indeed, if your threat model is honest-but-curious, even n-out-of-n secret sharing is fine. However, like vibs29 noted, that adversarial model is weaker. 

With point 2 mentioned by vibs29 I partly agree. For 8 participants, (kyber-interpretation-of)-t=3 gives the same protection as t=6. If I have this wrong, please show me how t=6 is weaker @vibs29.

**Comment by vibs29** [comment]
Hi there @seratym. Yes, thanks, I hope you are well too.

Yes, different papers use the same symbols to mean different things. But I am consistent everywhere with my symbols and everywhere I use (case-insensitive) T, it means the same thing (as I declared in the last sentence of my second post), and that thing matches Kyber's use of the symbol T. Further, I assumed that in this thread, everybody's use of T meant the same thing, i.e. the meaning in Kyber. (This discussion would be impossible if its participants used T to sometimes mean one thing and other times mean something else just because different papers used the same symbol for different things.) This thread (and Kyber) (should) have a single conceptual namespace in which T only ever means what Kyber means by it. I follow the discipline in this thread that when quoting a result from a paper that uses "T" to mean something else, I adjust and quote the result in terms of our (i.e. Kyber's) meaning of T. Reviewing this entire thread, you and I have been consistent everywhere by using T only to mean what Kyber means by it. So there has been no confusion between us yet. If we maintain this discipline, we'll continue to avoid this category of confusion entirely, as we've done so far.

Right, MinimumT doesn't make sense to me either, as I said. I also agree that there doesn't seem to be any point to having a function named MaximumT either. (To the extent that there is a function at all, it should be called MaximumT not MinimumT, which is what I realised in my second message in this thread. If you study the last link I posted (i.e. to issue 374), you may conclude that the existence of the function itself was misguided to begin with, sending us on a wild goose chase.)

Regarding n=8 and whether t=6 is weaker than t=3, I have no opinion. Once or twice in the past I've independently tried to understand why one might choose t=6 and failed. You and @ineiti have a better understanding than me about t=6 and curious versus byzantine adversaries etc.
